runtime/lua/vim/lsp/diagnostic.lua

```diff
@@ -57,338 +13,241 @@ end
-  [DiagnosticSeverity.Hint]        = { guifg = "LightGrey" };
-}
-
--- Make a map from DiagnosticSeverity -> Highlight Name
-local make_highlight_map = function(base_name)
-  local result = {}
-  for k, _ in pairs(diagnostic_severities) do
-    result[k] = "LspDiagnostics" .. base_name .. DiagnosticSeverity[k]
-  end
-
-  return result
-end
-
-local default_highlight_map = make_highlight_map("Default")
-local virtual_text_highlight_map = make_highlight_map("VirtualText")
-local underline_highlight_map = make_highlight_map("Underline")
-local floating_highlight_map = make_highlight_map("Floating")
-local sign_highlight_map = make_highlight_map("Sign")
-
--- }}}
--- Diagnostic Namespaces {{{
 local DEFAULT_CLIENT_ID = -1
-local get_client_id = function(client_id)
+---@private
+local function get_client_id(client_id)
   if client_id == nil then
     client_id = DEFAULT_CLIENT_ID
   end
 
   return client_id
 end
 
-local get_bufnr = function(bufnr)
+---@private
+local function get_bufnr(bufnr)
   if not bufnr then
-    return api.nvim_get_current_buf()
+    return vim.api.nvim_get_current_buf()
   elseif bufnr == 0 then
-    return api.nvim_get_current_buf()
+    return vim.api.nvim_get_current_buf()
   end
 
   return bufnr
 end
 
-
---- Create a namespace table, used to track a client's buffer local items
-local _make_namespace_table = function(namespace, api_namespace)
-  vim.validate { namespace = { namespace, 's' } }
-
-  return setmetatable({
-    [DEFAULT_CLIENT_ID] = api.nvim_create_namespace(namespace)
-  }, {
-    __index = function(t, client_id)
-      client_id = get_client_id(client_id)
-
-      if rawget(t, client_id) == nil then
-        local value = string.format("%s:%s", namespace, client_id)
-
-        if api_namespace then
-          value = api.nvim_create_namespace(value)
-        end
-
-        rawset(t, client_id, value)
-      end
-
-      return rawget(t, client_id)
-    end
-  })
+---@private
+local function severity_lsp_to_vim(severity)
+  if type(severity) == 'string' then
+    severity = vim.lsp.protocol.DiagnosticSeverity[severity]
+  end
+  return severity
 end
 
-local _diagnostic_namespaces = _make_namespace_table("vim_lsp_diagnostics", true)
-local _sign_namespaces = _make_namespace_table("vim_lsp_signs", false)
-
 ---@private
-function M._get_diagnostic_namespace(client_id)
-  return _diagnostic_namespaces[client_id]
+local function severity_vim_to_lsp(severity)
+  if type(severity) == 'string' then
+    severity = vim.diagnostic.severity[severity]
+  end
+  return severity
 end
 
 ---@private
-function M._get_sign_namespace(client_id)
-  return _sign_namespaces[client_id]
+local function diagnostic_lsp_to_vim(diagnostics)
+  return vim.tbl_map(function(diagnostic)
+    local start = diagnostic.range.start
+    local _end = diagnostic.range["end"]
+    return {
+      lnum = start.line,
+      col = start.character,
+      end_lnum = _end.line,
+      end_col = _end.character,
```

Comment:
Why does this need to read the filesystem? Don't we just need `vim.str_utfindex()`?

Reply:
Also, do you happen to have a simple test case for this that we can add to `lsp/diagnostic_spec.lua`? Right now there is no test for this behavior. If not, I'll try and whip something up.

Reply:
> Why does this need to read the filesystem? Don't we just need `vim.str_byteindex()`?

`str_byteindex` requires the line. For already loaded buffers it is possible to use `nvim_buf_get_lines`. For unloaded it would either be necessary to load them (reading vs + processing autocmds), or read the files directly, bypassing the autocmd processing. See https://github.com/neovim/neovim/pull/14615

Reply:
That makes sense. So if I'm understanding your suggestion here:

>An alternative would be to store the col format as part of the diagnostics and make the conversion to byte column lazy?

We don't need to really convert the byte columns until the point where diagnostics are actually displayed somehow (either through decorations or via the quickfix list). So we could indicate in some fashion to defer that conversion until the point of display.

I suppose we could support letting the `col` field of the diagnostic structure be a function, in which case it gets called in order to resolve the value.

This seems like a lot of additional complexity though and I'd want to be sure that the performance improvements are really worth it. So I'm leaning toward copying what folke did in #14615 first and then if we find that it's too slow we can try to be more clever.

Reply:
@mfussenegger Can you review 4e18d1ed5 and dce4259a8?

Reply:
Looks good :+1: 

> This seems like a lot of additional complexity though and I'd want to be sure that the performance improvements are really worth it. So I'm leaning toward copying what folke did in #14615 first and then if we find that it's too slow we can try to be more clever.

:100: